hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java

```diff
@@ -1336,7 +1341,7 @@ public HFileBlock readBlock(long dataBlockOffset, long onDiskBlockSize, final bo
             return cachedBlock;
           }
 
-          if (!useLock && cacheable && cacheConf.shouldLockOnCacheMiss(expectedBlockType)) {
+          if (!useLock && cacheBlock && cacheConf.shouldLockOnCacheMiss(expectedBlockType)) {
```

Comment:
Why we have this change?

Reply:
In HBASE-27474, we were only caching refs if compaction was off. Here we are modifying this behaviour. We'll let the cache implementation decide how to handle refs. For BucketCache, we are modifying it to first resolve the ref and check for the referred file block. If it's not cache, it will cache the ref as the block cache key.

Reply:
Same as above, better add some comments to explain the history~